src/uct/sm/cma/cma_ep.c

```diff
@@ -55,66 +55,37 @@ ucs_status_t uct_cma_ep_common_zcopy(uct_ep_h tl_ep,
                                                      unsigned long),
                                      char *fn_name)
 {
+    uct_cma_ep_t *ep     = ucs_derived_of(tl_ep, uct_cma_ep_t);
+    ssize_t delivered    = 0;
+    size_t local_iov_idx = 0;
     ssize_t ret;
-    ssize_t delivered = 0;
-    size_t iov_it;
-    size_t iov_it_length;
-    size_t iov_slice_length;
-    size_t iov_slice_delivered;
-    size_t local_iov_it;
-    size_t length = 0;
+    size_t local_iov_cnt;
+    size_t length;
     struct iovec local_iov[UCT_SM_MAX_IOV];
     struct iovec remote_iov;
-    uct_cma_ep_t *ep = ucs_derived_of(tl_ep, uct_cma_ep_t);
 
-    do {
-        iov_it_length = 0;
-        local_iov_it = 0;
-        for (iov_it = 0; iov_it < ucs_min(UCT_SM_MAX_IOV, iovcnt); ++iov_it) {
-            iov_slice_delivered = 0;
-
-            /* Get length of the particular iov element */
-            iov_slice_length = uct_iov_get_length(iov + iov_it);
-
-            /* Skip the iov element if no data */
-            if (!iov_slice_length) {
-                continue;
-            }
-            iov_it_length += iov_slice_length;
-
-            if (iov_it_length <= delivered) {
-                continue; /* Skip the iov element if transferred already */
-            } else {
-                /* Let's assume the iov element buffer can be delivered partially */
-                if ((iov_it_length - delivered) < iov_slice_length) {
-                    iov_slice_delivered = iov_slice_length - (iov_it_length - delivered);
-                }
-            }
-
-            local_iov[local_iov_it].iov_base = (void *)((char *)iov[iov_it].buffer +
-                                                        iov_slice_delivered);
-            local_iov[local_iov_it].iov_len  = iov_slice_length - iov_slice_delivered;
-            ++local_iov_it;
-        }
-        if (!delivered) {
-            length = iov_it_length; /* Keep total length of the iov buffers */
-        }
-
-        if(!length) {
-            return UCS_OK; /* Nothing to deliver */
-        }
+    local_iov_cnt = uct_iovec_fill_iov(local_iov, iov,
```

Comment:
do not you need to fill it by `UCT_SM_MAX_IOV` portions? (assuming UCT_SM_MAX_IOV < iov_cnt)

Reply:
good catch - fixed!

but looking in the previous code, it seems that the case when `iovcnt` > `UCT_SM_MAX_IOV ` didn't work